src/metabase/transforms/core.clj

```diff
@@ -162,54 +131,47 @@
     (assoc bindings name {:entity     (materialize/make-card-for-step! step query)
                           :dimensions (infer-resulting-dimensions local-bindings step query)})))
 
-(def ^:private ^{:arglists '([field])} field-type
-  (some-fn :special_type :base_type))
-
-(defn- satisfies-requierment?
-  [{requirement-dimensions :dimensions} table]
-  (let [table-dimensions (map field-type (:fields table))]
-    (every? (fn [dimension]
-              (some #(isa? % dimension) table-dimensions))
-            requirement-dimensions)))
-
-(defn- satisfy-requirements
+(s/defn ^:private satisfy-requirements :- Bindings
   [db-id schema {:keys [requires]}]
-  (let [tables   (table/with-fields
-                   (db/select 'Table :db_id db-id :schema schema))
-        bindings (m/map-vals (fn [requirement]
-                               (filter (partial satisfies-requierment? requirement) tables))
-                             requires)]
+  (let [tables   (for [table (table/with-fields
+                               (db/select 'Table :db_id db-id :schema schema))]
+                   ;; temporarily until we merge the sync PR
+                   (assoc table :domain_entity (de/domain-entity-for-table table)))
+        bindings (into {}
+                       (map (fn [requirement]
+                              (let [t (get-in @domain-entity-specs [requirement :type])]
+                                [requirement
+                                 (filter #(-> % :domain_entity :type (isa? t)) tables)])))
+                       requires)]
     ;; If multiple tables match punt for now
-    (when (every? (comp #{1} count second) bindings)
+    (when (every? (comp #{1} count val) bindings)
       (m/map-vals (fn [[table]]
                     {:entity     table
-                     :dimensions (into {} (for [field (:fields table)]
-                                            [(-> field field-type name) (mbql-reference field)]))})
+                     :dimensions (m/map-vals de/mbql-reference
+                                             (get-in table [:domain_entity :dimensions]))})
                   bindings))))
 
-(defn- store-requirements!
-  [db-id requirements]
+(s/defn ^:private store-requirements!
+  [db-id :- su/IntGreaterThanZero, bindings :- Bindings]
   (qp.store/fetch-and-store-database! db-id)
   (qp.store/fetch-and-store-fields!
-   (mapcat (comp (partial map u/get-id) :fields :entity val) requirements)))
+   (mapcat (comp (partial map u/get-id) :fields :entity val) bindings)))
 
 (s/defn apply-transform!
```

Comment:
I am really having trouble understanding what this function is supposed to be doing. It looks like it's just returning a sequence of IDs? But it's called `apply-transform!` 😕

Reply:
It's the IDs of newly created cards materializing given transform. I'm not 100% on the return value. It could be that it's more correct to return the full cards or something like that.